### PR TITLE
[FIX] marking of stored computed fields during creation

### PR DESCRIPTION
**Pull request by guewen**
Fixes 3a6ac95 that ignores computation of recursive stored fields during
creation of records.

ping @rco-odoo 

Description of the issue/feature this PR addresses:

The change in https://github.com/odoo/odoo/pull/63293 led to a regression observed in unit tests of an OCA module (https://github.com/OCA/stock-logistics-warehouse/pull/1035). The impacted field in the OCA module is https://github.com/OCA/stock-logistics-warehouse/blob/18f6c1e60adc7cb12c18868888bd1bcb3ba3f979/stock_vertical_lift/models/stock_location.py#L53-L64.

The problem happens when we have:

* a recursive stored computed field
* an `api.depends` of this field which is a relation to another record

When this "another record" is created and should recompute the recursive field, it is ignored.

Current behavior before PR:

On creation of records, dependent recursive fields are not recomputed.

Desired behavior after PR is merged:

On creation of records, dependent recursive fields are recomputed.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/models.py [hunk 1]
```diff
@@ -5831,19 +5831,19 @@ def modified(self, fnames, create=False, before=False):
...
                 else:
-                    if recursive:
+                    # Dont force the recomputation of compute fields which are
+                    # not stored as this is not really necessary.
+                    if not create and field.recursive:
                         marked_records = records & self.env.cache.get_records(records, field)
                     self.env.cache.invalidate([(field, records._ids)])
                 # recursively trigger recomputation of field's dependents
-                if recursive:
+                if field.recursive and marked_records:
```
Comment (rco-odoo) [inline odoo/models.py:5846]: The condition may be simplified to
```suggestion
                if marked_records:
```
You could even rename `marked_records` to `recursively_marked` to make things clearer.

**Review by rco-odoo — COMMENTED** [review]
Looks good to me.

However, the test models look more complex than necessary to me.
I suggest a simpler model that reproduces the issue:
```python
class ComputeRecursiveTree(models.Model):
    _name = 'test_new_api.recursive.tree'
    _description = 'Test New API Recursive with one2many field'

    name = fields.Char(required=True)
    parent_id = fields.Many2one('test_new_api.recursive.tree', ondelete='cascade')
    children_ids = fields.One2many('test_new_api.recursive.tree', 'parent_id')
    display_name = fields.Char(compute='_compute_display_name', store=True)

    @api.depends('name', 'children_ids.display_name')
    def _compute_display_name(self):
        for rec in self:
            children_names = rec.mapped('children_ids.display_name')
            rec.display_name = '%s(%s)' % (rec.name, ', '.join(children_names))
```
The test the consists in creating 3 records that form a 3-level tree:
```python
def test_12_recursive_tree(self):
    foo = self.env['test_new_api.recursive.tree'].create({'name': 'foo'})
    self.assertEqual(foo.display_name, 'foo()')
    bar = foo.create({'name': 'bar', 'parent_id': foo.id})
    self.assertEqual(foo.display_name, 'foo(bar())')
    baz = foo.create({'name': 'baz', 'parent_id': bar.id})
    self.assertEqual(foo.display_name, 'foo(bar(baz()))')
```

**Comment by guewen** [comment]
@rco-odoo Thanks, that's indeed better! I updated my commit

**Comment by guewen** [comment]
Amended to add `core:` in the commit summary

**Review by rco-odoo — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**